File: typescript/packages/core/tools/matimo_get_tool/matimo_get_tool.ts
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { validateToolDefinition, getGlobalMatimoLogger } from '@matimo/core';

interface GetToolParams {
  name: string;
  tool_dir?: string;
}

interface GetToolResult {
  found: boolean;
  name?: string;
  yaml_content?: string;
  definition?: Record<string, unknown>;
  message: string;
}

export default async function matimoGetTool(params: GetToolParams): Promise<GetToolResult> {
  const logger = getGlobalMatimoLogger();
  const toolDir = params.tool_dir ?? './matimo-tools';

  const defPath = path.join(toolDir, params.name, 'definition.yaml');
  if (!fs.existsSync(defPath)) {
    logger.warn('matimo_get_tool: tool not found', { name: params.name, path: defPath });
    return { found: false, message: `Tool "${params.name}" not found at ${defPath}` };
  }

  const yamlContent = fs.readFileSync(defPath, 'utf-8');

  let definition: Record<string, unknown>;
  try {
    const parsed = yaml.load(yamlContent);
    const validated = validateToolDefinition(parsed);
    // Omit internal _definitionPath from the returned object
    const { _definitionPath: _, ...rest } = validated as Record<string, unknown>;
    definition = rest;
  } catch (err) {
    return {
      found: true,
      name: params.name,
      yaml_content: yamlContent,
      message: `Tool YAML is invalid: ${(err as Error).message}`,
    };
  }

  logger.debug('matimo_get_tool: retrieved', { name: params.name });

  return {
    found: true,
    name: params.name,
    yaml_content: yamlContent,
    definition,
    message: `Tool "${params.name}" retrieved successfully`,
  };
}

File: python/packages/core/src/matimo/tools/matimo_get_tool/matimo_get_tool.py
"""matimo_get_tool — retrieve the full definition of a tool (YAML + parsed fields)."""
from __future__ import annotations

import logging
import re
from pathlib import Path

import yaml

logger = logging.getLogger("matimo")

UNSAFE_NAME = re.compile(r"[/\\]|\.\.|[\x00-\x1f]")


async def run(params: dict) -> dict:  # type: ignore[type-arg]
    from matimo.core.models import ToolDefinition

    name: str = params.get("name", "")
    tool_dir: str = params.get("tool_dir", "./matimo-tools")

    if not name or not name.strip():
        return {"found": False, "message": "Tool name is required"}
    if UNSAFE_NAME.search(name):
        return {
            "found": False,
            "message": "Tool name contains invalid characters (path traversal, backslash, or control characters)",
        }

    def_path = Path(tool_dir) / name / "definition.yaml"
    if not def_path.exists():
        logger.warning("matimo_get_tool: tool not found at %s", def_path)
        return {"found": False, "message": f'Tool "{name}" not found at {def_path}'}

    yaml_content = def_path.read_text(encoding="utf-8")

    try:
        parsed = yaml.safe_load(yaml_content)
        tool_def = ToolDefinition.model_validate(parsed)
    except Exception as exc:
        return {
            "found": True,
            "name": name,
            "yaml_content": yaml_content,
            "message": f"Tool YAML is invalid: {exc}",
        }

    # Convert to dict, excluding internal private fields
    definition = tool_def.model_dump(exclude_none=True)
    definition.pop("_definition_path", None)

    logger.debug("matimo_get_tool: retrieved %s", name)

    return {
        "found": True,
        "name": name,
        "yaml_content": yaml_content,
        "definition": definition,
        "message": f'Tool "{name}" retrieved successfully',
    }
